### ISSUE TITLE
Browser (oauth2) auth method

### ISSUE THREAD

**Issue by andrewbaxter** [jmgilman/vaultrs#101]
I'm not sure if this would be out of scope, but it would be very convenient to have it here rather than make a whole new crate just for browser auth.

When using hashicorp cloud for local tools/automation, their official instructions say you need to create a service principal to generate short-lived api tokens.  When dealing with a team, that means you need to share the service principal credentials, defeating the purpose of using vault to begin with.

The hcp client itself doesn't do this obviously.  It uses (undocumented?) oauth2 with the standard browser flow.  It doesn't look too hard to implement, and it would provide a way for users to be more secure and manage fewer credentials at the same time.

Just in case you're unfamiliar with it, the method is pretty simple:
- Start an http server on localhost (127.0.0.1)
- Open the login link in the browser.  Typically this is either printed to the console or automatically opened in a browser via the platform-specific run method (`open` on mac I think, `xdg-open` on linux, not sure about windows). The user completes the login flow.
- After logging in, the browser posts the payload to the local server
- The local server exits and exchanges the post payload for an actual token

Generally speaking to get this to work you need to make sure the parameters match what hcp uses: The redirect address and port you use match exactly with what `hcp` uses (so if hcp uses `localhost` you need to use `localhost` and not `127.0.0.1`), but also the audience, etc.

For reference, the code the `hcp` uses is here:
- https://github.com/hashicorp/hcp-sdk-go/blob/main/auth/browser.go - setting up the server
- https://github.com/hashicorp/hcp-sdk-go/blob/main/config/new.go - oauth2 params

I'd be happy to try implementing it, but I'd need to know what http server to use (or just plain hyper) and any other design considerations... also maybe it'd make sense to publish the oauth2 params in case someone wants to use it in another environment? Like I guess for a service that integrates with vault?

**Comment by andrewbaxter** [comment]
FWIW there's precedence for stuff like this in SDKs. The Azure go sdk contains functions for browser login.  I think it stores the credentials in the default location for the user.

**Comment by Haennetz** [comment]
Hey @andrewbaxter, did I assume that right that you plan to use this create for accessing a HCP vault instance?

**Comment by andrewbaxter** [comment]
Yeah, precisely.  Although the more I use it the more it seems different from the self hosted vault I'm used to...

**Comment by Haennetz** [comment]
I also found a [documentation](https://developer.hashicorp.com/hcp/docs/hcp/api) about autenticate to HCP.  If you want fell free to add it you can use the new modul named hcp for that. I don't have any expierence with the HCP.
